server/aws-lsp-identity/src/sharedConfig/unmergeConfigFiles.ts

```diff
@@ -0,0 +1,106 @@
+// it is fixed with a focus on retaining as much of the customer's file as reasonable, applying best practices
+// in cases where security is a concern.
+function unmergeSectionToConfigFiles(
+    sectionName: string,
+    mergedSettings: IniSection,
+    configFile: ParsedIniData,
+    credentialsFile: ParsedIniData
+): void {
+    const configSection = (configFile[sectionName] ||= {})
+    const credentialsSection = (credentialsFile[sectionName] ||= {})
+
+    // Remove existing settings not on mergedSettings (i.e. deleted)
+    removeDeletedEntries({ removeFrom: configSection, deletedFrom: mergedSettings })
+    removeDeletedEntries({ removeFrom: credentialsSection, deletedFrom: mergedSettings })
+
+    // Apply each setting to the correct file section, in some cases, both
+    for (const [settingName, settingValue] of Object.entries(mergedSettings)) {
+        const inConfig: boolean = configSection && Object.hasOwn(configSection, settingName)
+        const inCredentials: boolean = credentialsSection && Object.hasOwn(credentialsSection, settingName)
+
+        // BEST PRACTICE: Secrets should be stored in credentials only
+        // https://docs.aws.amazon.com/sdkref/latest/guide/file-format.html#file-format-creds
+        if (isSecretSetting(settingName)) {
+            credentialsSection[settingName] = settingValue
+
+            if (inConfig) {
+                delete configSection[settingName]
+            }
+            continue
+        }
+
+        // Whether non-secret setting wasn't previously stored or just in config, prefer config
+        if (!inCredentials) {
+            configSection[settingName] = settingValue
+            continue
+        }
+
+        // Otherwise set in credentials and only update in config if it exists
+        credentialsSection[settingName] = settingValue
+        inConfig && (configSection[settingName] = settingValue)
+    }
+
+    // Remove empty sections
+    !Object.keys(configSection).length && delete configFile[sectionName]
+    !Object.keys(credentialsSection).length && delete credentialsFile[sectionName]
+}
+
+export function removeDeletedEntries(args: { removeFrom: Record<string, unknown>; deletedFrom: object }): void {
+    const { removeFrom, deletedFrom } = args
+
+    if (!(removeFrom && deletedFrom)) {
+        return
+    }
+
+    for (const key of Object.keys(removeFrom)) {
+        if (!Object.hasOwn(deletedFrom, key)) {
+            delete removeFrom[key]
+        }
+    }
+}
+
+function isSecretSetting(name: string): boolean {
+    // Settings containing "secrets" as defined in the public docs are to be stored in creds
+    // if not already defined in config (this is for "retain" support, future options may
+    // allow forcing secrets to always be written to credentials, existing or not).
+    // https://docs.aws.amazon.com/sdkref/latest/guide/file-format.html#file-format-creds
+    return ['aws_access_key_id', 'aws_secret_access_key', 'aws_session_token'].includes(name)
```

Comment:
Are they going to be always lowercase?

Reply:
Good catch, while all settings are lowercase in public documents, none state whether the files are case-sensitive or not.  Seems like this has caused some questions/problems in the past (see links below).  It appears that the AWS CLI is case-insensitive, but most AWS SDKs are case-sensitive.

Made changes to take the approach of being able to read any setting name regardless of case, but when writing, setting names are converted to lowercase to make coding in TypeScript easier (as property names are case-sensitive) and to promote better compatibility with AWS SDKs on file saves.

https://github.com/aws/aws-toolkit-vscode/issues/2394
https://github.com/aws/aws-sdk-ruby/issues/2458
https://github.com/aws/aws-sdk/issues/574
https://github.com/awslabs/aws-sdk-rust/issues/1020